### ISSUE TITLE
Very slow tests since vesion 12.2.0

### ISSUE THREAD

**Issue by lousley** [ike18t/ng-mocks#1452]
Hello together

Since we updated ng-mocks to version 12.2.0 on our project (Angular 12 with unittests jasmine / karma), we are seeing extreme performance issues. Do you have any spontaneous idea what this could be?
I did some debugging and saw that defineTouches / generateTouches fills the touches SET with extremely many entries.

We only use in some tests ngMocks and MockComponent.

Thanks for any feedback.

**Comment by satanTime** [comment]
Hi @lousley , might I ask you to share a test which demonstrates the degradation?

**Comment by satanTime** [comment]
Also, to confirm - 12.5.0 also has the issue, right?

**Comment by lousley** [comment]
yes i have the problem also with version 12.5. interestingly i just saw that if i have a testfile without ngmocks imports or functionalities but i have many imports and providers it becomes extremely slow. defineTouches and geneateTouches is probably the problem even if no ng-mocks is needed in the corresponding file?

**Comment by lousley** [comment]
performance profile:
<img width="1028" alt="grafik" src="https://user-images.githubusercontent.com/20418276/146172565-39a21c6c-2296-4c10-bdcc-9d05e9769606.png">

**Comment by lousley** [comment]
is there any way to turn off/override this generatetouches?

**Comment by LPCmedia** [comment]
seeing the same thing in ng 13.0.2  + ng-mocks 12.5.0. Performance is especially slow when MockModule is added. 


| Same Unit Test using:                     | Seconds |
| ----------------------------------------- | ------- |
| Mock Module + MockComponent               | 38.351  |
| MockComponents                            |   6.999  |
| ***Benchmarks:***                         |         |
| NO_ERRORS_SCHEMA + CUSTOM_ELEMENTS_SCHEMA |   3.742s |
| using actual modules                      |   6.558s |

with ( MockModule + MockComponent) 
![image](https://user-images.githubusercontent.com/5975736/146292474-53488f40-ddc4-464b-98eb-87a447c51693.png)

with (MockComponents)

![image](https://user-images.githubusercontent.com/5975736/146293539-716774e1-dfb4-40ec-8c33-12b5c6693bac.png)

without: 

![image](https://user-images.githubusercontent.com/5975736/146292565-1c924052-8469-4eba-82d2-b4b234b21d87.png)

using the actual modules: 

![image](https://user-images.githubusercontent.com/5975736/146292936-5e6aaf1c-08f0-4cd7-aed4-0c664825ef10.png)

it's probably not a fair comparison with `NO_ERRORS_SCHEMA` something is off since ng13.

**Comment by satanTime** [comment]
Hi there, thanks for the info and support.

might you also downgrade `ng-mocks` to `<12.2.0` and share results?

looks like another change in A13 will cause huge refactoring, so I could try to find the issue which impacts performance degradation and to fix it there too.

**Comment by satanTime** [comment]
Regarding `MockModule`, it depends in the amount of declarations recursively. Because it goes through the whole declaration tree of the module and mocks every thing. So to mock `AppModule` will take much more time than `ButtonModule`.
Another issue here is jest, because it runs tests in isolated runtimes, it bootstraps every runtime from scratch and causes mocking for the same things. That also kills performance, but it's a separate story.

Considering that a simple test takes 6.5 seconds on your side, I would say that your app looks heavy and can have a lot of declarations, and that's why `ng-mocks` needs enormous time to mock all of them.

I'll try to find a way to add metrics so we could debug where `ng-mocks` looses time.

**Comment by LPCmedia** [comment]
sounds like a good plan on the metrics. 

As a guideline we will try to avoid MockModule, we are supposed to be shallow testing anyway 🤷 . But the atomic option is to appealing. 

Is there a separate issue logged for jest performance improvements ?

**Comment by lousley** [comment]
In my case i have the following TestBed

<img width="889" alt="grafik" src="https://user-images.githubusercontent.com/20418276/146323754-f55d3848-e52c-4056-9d61-dae91ec09d3f.png">

I think the problem is the AwbTestingModule with many declarations / imports / exports and providers as:

<img width="873" alt="grafik" src="https://user-images.githubusercontent.com/20418276/146323868-7cf58696-8256-4944-9c47-fb6fe9121297.png">

If I remove the TestingModule and use only the required declarations / imports / providers in the corresponding test file, then the test is already much faster. I also see that in this testfile no ng-mocks stuff is needed at all, nevertheless it is very slow.
We use this TestModule in quite a few places, because it is easier and more convenient than defining the same things over and over again. 
Is there a way to tell ng-mocks to ignore this module when generatingTouches / definingTouches? Or am I looking at this wrong and the problem is somewhere else?

**Comment by LPCmedia** [comment]
> Hi there, thanks for the info and support.
> 
> might you also downgrade `ng-mocks` to `<12.2.0` and share results?
> 
> looks like another change in A13 will cause huge refactoring, so I could try to find the issue which impacts performance degradation and to fix it there too.

can you confirm : ng-mocks 12.1.2 with ng13 ?

**Comment by gagle** [comment]
I can also confirm that. We are using Angular 12 + ng-mocks 12.1.0 and any version above 12.1.0 is extremely slowing down tests.

**Comment by satanTime** [comment]
Hey @gagle, might you provide how much time it needed for the test above?

**Comment by lousley** [comment]
@satanTime: i would also like to share my time difference:

Version 12.1.2: 277.31s
Version >12.1.2: 1200s

This based on 5500 tests

**Comment by satanTime** [comment]
thanks for the numbers, ill try to figure out how to optimise that.

**Comment by lousley** [comment]
any updates? :S

**Comment by satanTime** [comment]
Hi,

not much, jest doesn't want to merge the fix which would simplify the flow. And I haven't had time yet to dig deeper and to find alternative solutions.

Also, feel free to post any findings here. For example, if you see that some functions cause the issue and others aren't. It would really help once I'm on fixing it.

**Comment by rbirkgit** [comment]
I cannot get our tests to even start with after updating to angular 13 and ng-mocks v13. I tried downgrading to 12.x but that doesn't seem compatible with ang13. 

I start the unit tests with `yarn test` and after compilation the chrome browser starts up and then just hangs. Only after I remove all references to ng-mocks I can get the tests running again. In our ang12 version all worked fine.

**Comment by satanTime** [comment]
Hi @rbirkgit,

Thanks for the info. Looks like there is a recursion.
Would it be possible to share your project with me for investigation?

**Comment by rbirkgit** [comment]
Unfortunately it's a big project with many dependencies and don't think we can share the code. The strange thing is, removing all references to ng-mocks all is fine (just to get ta base line). Then all I do is adding one place with MockComponent(xxx) and boom, it hangs using v13.

Not sure if there are some ways to get more debug info or how to test what could cause the issue?

**Comment by rbirkgit** [comment]
All I do is add this in one of our tests and the whole thing hangs before even the first test starts.

```
  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [MockComponent(FileInputComponent)],
```

**Comment by satanTime** [comment]
Yep, I think some root installs don't work well. could you check for example not `MockComponent`, but `MockRender` or `ngMocks.*`?

I'll prepare a special debug version which should shed light on the source of the issue, would it be possible for you to run it and share with me its output?

**Comment by rbirkgit** [comment]
I would be happy to test and run debug versions and share results.

**Comment by rbirkgit** [comment]
I tried creating a simple angular-cli project with a dummy component to mock. All worked there. Just to make sure my env is all fine.

Then removed all spec files in our project (~250). Left one with the MockComponent. Have same issue as before. Then tried mocking a dummy empty component to see if it's the component itself, but now keep getting this error.

`
Error: Unexpected directive 'MockOfMyTestComponent' imported by the module 'DynamicTestModule'. Please add an @NgModule annotation.
`

Need to figure out what I am doing wrong here. May need to delete a lot of code line by line until it starts working, but many lines of code and depending libs. One step at a time.

**Comment by rbirkgit** [comment]
Played more today and my latest conclusion is that it's not actually hanging, it really really slow. We have a 60s timeout where it will disconnect the test and that what I am experiencing. And the browser then just sits there.

So we have a shared MockModule where we have mocked the most common services etc we use. Then I went ahead and deleted every spec file we have except one, which is a bare minimum. Just test that component is created. 

Now I added a dummy empty component and added MockComponent(MyTestComponent) to this minimal test. Still hangs, or rather timeouts ater 60sec.

So started removing mocks in our mock module and after removing enough of them, the app.component test passed taking 25s. It usually is much faster, much less than 1sec. When we add the MockComponent things explode exponentially.

**Comment by satanTime** [comment]
Do you mean 25s without ng-mocks? Or without it it's 1s?

**Comment by rbirkgit** [comment]
The test takes 0.3sec normally. The moment I add a single line mocking a dummy empty component, it takes 25.4sec

`MockComponent(MyTestComponent)`

**Comment by rbirkgit** [comment]
This is my code in my test:

```
  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [AppComponent, MockComponent(MyTestComponent)],
      imports: [MockModule]
```

I had to delete mocks in our MockModule one by one until the test got under the 60sec timeout. Not sure why code in our MockModule should affect anything when adding the `MockComponent(xxx)` call.

**Comment by lousley** [comment]
we had the same problem, but since we upgraded to angular 13.2.4, this problem is no longer so extreme. What version are you running and what does your test.ts file look like?

**Comment by rbirkgit** [comment]
We are using Angular 13.3.0

test.ts is the default one:
```
declare const require: {
  context(
    path: string,
    deep?: boolean,
    filter?: RegExp
  ): {
    <T>(id: string): T;
    keys(): string[];
  };
};

getTestBed().initTestEnvironment(BrowserDynamicTestingModule, platformBrowserDynamicTesting());
const context = require.context('./', true, /\.spec\.ts$/);
context.keys().map(context);
```

**Comment by lousley** [comment]
can you test with thi initTesEnvironment:

`getTestBed().initTestEnvironment(
  BrowserDynamicTestingModule,
  platformBrowserDynamicTesting(), {
    teardown: { destroyAfterEach: false }
}
);`

**Comment by rbirkgit** [comment]
Tested turning off teardown. No difference. Then right now I literally only running one single test that all it does it to test the component got created and adding the mockComponent(MyTestComponent).

**Comment by rbirkgit** [comment]
Here are the dependencies:
```
"dependencies": {
  "@angular/animations": "~13.3.0",
  "@angular/common": "~13.3.0",
  "@angular/compiler": "~13.3.0",
  "@angular/core": "~13.3.0",
  "@angular/forms": "~13.3.0",
  "@angular/platform-browser": "~13.3.0",
  "@angular/platform-browser-dynamic": "~13.3.0",
  "@angular/router": "~13.3.0",
  "@auth0/angular-jwt": "~5.0.2",
  "@cds/core": "~5.6.4",
  "@clr/angular": "~13.0.3",
  "@clr/ui": "~13.0.3",
  "@fingerprintjs/fingerprintjs": "~3.3.3",
  "@jsplumb/browser-ui": "~5.5.2",
  "@jsplumb/connector-flowchart": "~5.5.2",
  "ace-builds": "~1.4.14",
  "bowser": "~2.11.0",
  "file-loader": "~6.2.0",
  "highcharts": "~9.3.3",
  "iframe-resizer": "~4.3.2",
  "ip-address": "^6.4.0",
  "ip-cidr": "^2.1.5",
  "js-sha256": "~0.9.0",
  "moment": "~2.29.1",
  "ngx-infinite-scroll": "~10.0.1",
  "node-cidr": "~0.1.8",
  "protractor-fail-fast": "3.1.0",
  "rxjs": "~7.5.5",
  "semver": "^7.3.5",
  "text-diff": "1.0.1",
  "tslib": "^2.3.1",
  "tus-js-client": "~2.3.1",
  "zone.js": "~0.11.5"
},
"devDependencies": {
  "@angular-devkit/build-angular": "~13.3.0",
  "@angular-eslint/builder": "^12.7.0",
  "@angular-eslint/eslint-plugin": "^12.7.0",
  "@angular-eslint/eslint-plugin-template": "^12.7.0",
  "@angular-eslint/schematics": "^12.7.0",
  "@angular-eslint/template-parser": "^12.7.0",
  "@angular/cdk": "~13.3.0",
  "@angular/cli": "~13.3.0",
  "@angular/compiler-cli": "~13.3.0",
  "@angular/language-service": "~13.3.0",
  "@types/file-saver": "~2.0.5",
  "@types/follow-redirects": "^1.14.1",
  "@types/ip-address": "~5.8.2",
  "@types/jasmine": "~4.0.0",
  "@types/js-yaml": "^4.0.5",
  "@types/node": "~14.18.12",
  "@types/properties-reader": "0.0.2",
  "@types/semver": "^7.3.9",
  "@types/tus-js-client": "~2.1.0",
  "@types/uuid": "~8.3.4",
  "@typescript-eslint/eslint-plugin": "^4.33.0",
  "@typescript-eslint/parser": "^4.33.0",
  "chrome-har": "^0.13.0",
  "dompurify": "~2.3.6",
  "eslint": "^7.32.0",
  "eslint-config-prettier": "^8.5.0",
  "eslint-plugin-jasmine": "^4.1.3"
  "file-saver": "~2.0.5",
  "follow-redirects": "^1.14.9",
  "jasmine": "~4.0.2"
  "jasmine-core": "~4.0.1",
  "jasmine-reporters": "~2.5.0",
  "jasmine-spec-reporter": "~7.0.0",
  "js-yaml": "~4.1.0",
  "karma": "~6.3.17",
  "lodash-es": "~4.17.21",
  "lottie-web": "~5.9.1",
  "lz-string": "~1.4.4",
  "ng-mocks": "~13.2.0",
  "ng-packagr": "^13.3.0",
  "ngx-lottie": "~8.0.1",
  "prettier": "2.5.1",
  "properties-reader": "0.3.1",
  "protractor": "~7.0.0",
  "protractor-jasmine2-screenshot-reporter": "^0.5.0",
  "protractor-video-reporter": "^0.3.0",
  "ts-node": "~10.7.0",
  "typescript": "~4.6.2",
  "webpack-bundle-analyzer": "~4.5.0"
},
```

**Comment by satanTime** [comment]
Hi @rbirkgit,

could give me a favor and test the next conditions:

preconditions: you have one spec file which needs 25 secs to be run.
in its test, please, remove `MockComponent` and simply pass its component into `TestBed` like a test without `ng-mocks`.
Before the very first `describe` in this file, please add `ngMocks.autoSpy('jasmine');`.
The idea is that `ng-mocks` has been imported, but isn't used directly.

Now, please check how much time is needed to execute the test.
if it needs more than 1 sec, please keep `beforeEach` as it is, but delete all `it`.
Now add one `it` like that:

```ts
it('test', () => {
  expect(1).toEqual(1);
});
```

Please check how much time it needs.

after that, please add the same `it` again, so there would be 2 of them.

Please check how much time it needs.

---

At the end, you should have 3 numbers:
- how much test needs with one import of `ng-mocks`
- how much `beforeEach` consumes with one simple test
- how much `beforeEach` consumes with two simple test

**Comment by rbirkgit** [comment]
I did the first step as you described. I removed all spec files except one simple one with one it that only tests for component creation (default test). This runs in less than a second. Added the line you described above the describe:

```
import { ngMocks } from 'ng-mocks';

ngMocks.autoSpy('jasmine');

describe('AppComponent', () => {
...
```

Now tests disconnects after 60s (timeout). Without the ngMocks like, again it's less than a second.

**Comment by satanTime** [comment]
And to confirm: MockModule is super fat, isn't it?

**Comment by rbirkgit** [comment]
Yes, our MockModule is fat. I started removing code from it and it got faster, but still very slow. Removed enough, I got it to 25sec. Removed most, 15sec. We also have what may or may not be a separate issue that after the test, it still hangs. Meaning it says the it() passed but testing is still sitting there not exiting. Removing ng-mocks, it exits right away.

I know this is not much to go on but thought I try to describe. Our tests could be optimized with mocking things more properly. Right now it's not mocking a component's every dependent service and some are running during the tests, which using other services, along with observables etc. That where we wanted to cleanup the tests with ng-mocks. It was very promising and worked great in the ang12 env. Thanks for looking into it. Would be awesome if we can continue using it. It's a very nice tool.

**Comment by rbirkgit** [comment]
Tests both options. No difference.

**Comment by satanTime** [comment]
I was able to reproduce the issue. Working on a fix.

**Comment by satanTime** [comment]
Hi all,

could you verify that this build solves the issue?
simply put ng-mocks from the archive to node_modules of your project: [ng-mocks.zip](https://github.com/ike18t/ng-mocks/files/8355287/ng-mocks.zip).

**Comment by rbirkgit** [comment]
I will try when I get back to my computer later today. Thanks!!

**Comment by rbirkgit** [comment]
Tested it. Success!! It's as fast as v12 now and works! I need to pinch myself and try again :)

**Comment by satanTime** [comment]
Awesome!!! Thanks for confirming. I'll release it tomorrow.

**Comment by rbirkgit** [comment]
Now all tests run and normal speed. Got 6 tests failing of 350 in this project. Need to figure out next if it's a new issue or we did something wrong that now got exposed. Four them has `MockModule(SDKModule)` which is one of our modules that contains some services etc. It's causing:

```
NullInjectorError: R3InjectorError(DynamicTestModule)[ApplicationModule -> ApplicationRef -> ApplicationInitStatus -> InjectionToken Application Initializer -> [object Object] -> TimeoutService -> SessionService -> HttpClient -> HttpClient]: 
NullInjectorError: No provider for HttpClient!
```

This was working before. I'm not sure yet if it's an ang13 update issue, or a new ngMocks problem or our code just bad. But thought I post here just in case. Will look at it more on Monday.

**Comment by satanTime** [comment]
Why do you not want to use MockBuilder? It resolves the hell of dependencies. 

Anyway, if there is a bug, please create a new issue.

**Comment by satanTime** [comment]
Hi @lousley, could you also check whether it works for you?

And I would like to thank your for helping with debug information. It was very useful.

**Comment by lousley** [comment]
HI @satanTime i can confirm that the performance is back as before. thanks for the fix. really nice!

**Comment by satanTime** [comment]
[v13.3.0](https://github.com/ike18t/ng-mocks/releases/tag/v13.3.0) has been released and contains a fix for the issue. Feel free to reopen the issue or to submit a new one if you meet any problems.
